File: LeetCode/Study Plan/Practice Problems/Medium/The Maze/Go/the_maze.go
package main

import "fmt"

//Note the ball is rolling
func has_path(maze [][]int, start []int, dst []int) bool {
	n, m := len(maze), len(maze[0])
	visited := make([][]bool, n)
	for i := 0; i < n; i++ {
		visited[i] = make([]bool, m)
	}
	return _has_path(maze, start[0], start[1], dst[0], dst[1], n, m, &visited)
}

func _has_path(maze [][]int, i, j, x, y, n, m int, visited *[][]bool) bool {
	if (*visited)[i][j] {
		return false
	}
	if i == x && j == y {
		return true
	}
	(*visited)[i][j] = true
	r, l, u, d := j+1, j-1, i-1, i+1
	for r < m && maze[i][r] == 0 {
		r++
	}
	if _has_path(maze, i, r-1, x, y, n, m, visited) {
		return true
	}

	for l >= 0 && maze[i][l] == 0 {
		l--
	}
	if _has_path(maze, i, l+1, x, y, n, m, visited) {
		return true
	}

	for u >= 0 && maze[u][j] == 0 {
		u--
	}
	if _has_path(maze, u+1, j, x, y, n, m, visited) {
		return true
	}

	for d < n && maze[d][j] == 0 {
		d++
	}
	if _has_path(maze, d-1, j, x, y, n, m, visited) {
		return true
	}
	return false
}

func main() {
	fmt.Println(has_path([][]int{
		{0, 0, 1, 0, 0}, {0, 0, 0, 0, 0}, {0, 0, 0, 1, 0}, {1, 1, 0, 1, 1}, {0, 0, 0, 0, 0},
	}, []int{0, 4}, []int{4, 4}))
}

File: LeetCode/Medium/The Maze/Java/theMaze.java
class Solution {
    int n, m;
    int[][] dir = {{0, -1}, {1, 0}, {0, 1}, {-1, 0}};
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        n = maze.length;
        m = maze[0].length;
        boolean[][] visited = new boolean[n][m];
        return dfs(maze, start, destination, visited);
    }

    private boolean dfs(int[][] maze, int[] start, int[] dst, boolean[][] visited) {
        if (visited[start[0]][start[1]]) return false;
        if (start[0] == dst[0] && start[1] == dst[1]) return true;
        visited[start[0]][start[1]] = true;
        for (int i=0; i<4; i++) {
            int r = start[0], c = start[1];
            while(r >=0 && c >=0 && r < n && c < m && maze[r][c]==0) {
                r += dir[i][0];
                c += dir[i][1];
            }
            if (dfs(maze, new int[]{r-dir[i][0], c-dir[i][1]}, dst, visited)) return true;
        }
        return false;
    }
}
